### DISCUSSION TITLE
Firefly Importer tool timing out while trying to get a token during initial setup

### DISCUSSION THREAD

**Discussion by ghost** [firefly-iii/firefly-iii#7865] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I get a Server Error 500 in the browser and a cURL error 28 in the logs when the Firefly Importer tool is trying to connect back to the Firefly app. I am trying to run a fresh install of Firefly-iii on a new server, with new docker-compose.yml, .env, .db.env, and .importer.env files. I have made the necessary changes to the .env files per the Docker installation instructions. (DB_PASSWORD, MYSQL_PASSWORD, FIREFLY_III_URL).

### Debug information

Debug information generated at 2023-08-18 16:05:58 America/Denver for Firefly III version **6.0.19**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.19 |
| Firefly III API | 2.0.4 |
| Build | 553, 28-07-2023 14:22:29 UTC |
| Base Build | 13, 26-07-2023 07:34:31 CEST |
| DB version | 19 (exp. 19) |
| PHP | `8.2.8` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | America/Denver |
| Browser TZ | America/Denver |
| App environment | local |
| App debug mode | true |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |

### Expected behaviour

I haven't used the Importer tool before, but I would expect it to connect to Firefly and allow me to start importing CSV files.

### Steps to reproduce

1. Navigate to http://192.168.1.30:91 which redirects to http://192.168.1.30:91/token
2. Type http://192.168.1.30:90 into the "Firefly III URL" box
3. Type 3 into the "Client ID" box (Which I created in Firefly -> Options -> Profile -> OAuth -> Create New Client. I did make sure to uncheck "Confidential". The redirect URL is http://192.168.1.30:91/callback)
4. Click "Authorize"
5. Receive 500 | Server Error

### Additional info

Due to port conflicts, Firefly-iii runs on port 90:8080 and the Importer runs on 91:8080.

$ docker compose up -d
$ docker logs firefly_iii_importer -f
Now in entrypoint.sh for Firefly III Data Importer
Script:        1.6 (2022-02-17)
User:          'root'
Group:         'root'
Working dir:   '/var/www/html'
Build number:  234
Build date:    16-07-2023 04:28:08 UTC
Now parsing _FILE variables.
done!
Firefly III data importer v1.3.3
PHP: cli 8.2.8 Linux
Will now run Apache web server:
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 192.168.224.5. Set the 'ServerName' directive globally to suppress this message
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 192.168.224.5. Set the 'ServerName' directive globally to suppress this message
[Fri Aug 18 16:22:16.561200 2023] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.56 (Debian) configured -- resuming normal operations
[Fri Aug 18 16:22:16.561229 2023] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
[2023-08-18 16:22:38] local.DEBUG: Now in App\Http\Controllers\IndexController::index  
[2023-08-18 16:22:38] local.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
[2023-08-18 16:22:38] local.DEBUG: No valid secrets, redirect to token.index  
192.168.1.238 - - [18/Aug/2023:16:22:38 -0600] "GET / HTTP/1.1" 302 1539 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"
[2023-08-18 16:22:38] local.DEBUG: Now at App\Http\Controllers\TokenController::index  
[2023-08-18 16:22:38] local.DEBUG: No access token in hasAccessToken(), will return config variable.  
[2023-08-18 16:22:38] local.DEBUG: No client id in hasClientId(), will return config variable.  
[2023-08-18 16:22:38] local.DEBUG: No base url in getBaseUrl(), will return config variable.  
[2023-08-18 16:22:38] local.DEBUG: No vanity url in getVanityUrl(), will return config variable.  
[2023-08-18 16:22:38] local.INFO: The following configuration information was found:  
[2023-08-18 16:22:38] local.INFO: Personal Access Token: "" (limited to 25 chars if present)  
[2023-08-18 16:22:38] local.INFO: Client ID            : "0"  
[2023-08-18 16:22:38] local.INFO: Base URL             : "http://192.168.1.30:90"  
[2023-08-18 16:22:38] local.INFO: Vanity URL           : "http://192.168.1.30:90"  
192.168.1.238 - - [18/Aug/2023:16:22:38 -0600] "GET /token HTTP/1.1" 200 2771 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"
192.168.1.238 - - [18/Aug/2023:16:22:38 -0600] "GET /fa/css/all.min.css HTTP/1.1" 200 13028 "http://192.168.1.30:91/token" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"
192.168.1.238 - - [18/Aug/2023:16:22:38 -0600] "GET /css/bootstrap-dark.min.css?version=1.3.3 HTTP/1.1" 200 31464 "http://192.168.1.30:91/token" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"
[2023-08-18 16:22:45] local.DEBUG: Now at App\Http\Controllers\TokenController::submitClientId  
[2023-08-18 16:22:45] local.DEBUG: Submitted data:  {"client_id":"3","base_url":"http://192.168.1.30:90"} 
[2023-08-18 16:22:45] local.DEBUG: Base URL is "http://192.168.1.30:90"  
[2023-08-18 16:22:45] local.DEBUG: Vanity URL is now "http://192.168.1.30:90"  
[2023-08-18 16:22:45] local.DEBUG: Base URL is now "http://192.168.1.30:90"  
[2023-08-18 16:22:45] local.DEBUG: Now in App\Http\Controllers\TokenController::redirectForPermission(request, "http://192.168.1.30:90", "http://192.168.1.30:90", 3)  
[2023-08-18 16:22:45] local.DEBUG: Query parameters are {"client_id":3,"redirect_uri":"http://192.168.1.30:91/callback","response_type":"code","scope":"","state":"FrAHwauL24rOq2QHpToWt5HX6puqNNKrCIYRbR7d","code_challenge":"GpQ-od8TfSsEyIPaP77grn3bOO0lSmt92clbClvb864","code_challenge_method":"S256"} 
[2023-08-18 16:22:45] local.DEBUG: Now redirecting to "http://192.168.1.30:90/oauth/authorize?" (params omitted)  
192.168.1.238 - - [18/Aug/2023:16:22:45 -0600] "POST /token/client_id HTTP/1.1" 302 2815 "http://192.168.1.30:91/token" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"
[2023-08-18 16:22:49] local.DEBUG: Now at App\Http\Controllers\TokenController::callback  
[2023-08-18 16:22:49] local.DEBUG: State is valid!  
[2023-08-18 16:22:49] local.DEBUG: Params for access token {"form_params":{"grant_type":"authorization_code","client_id":3,"redirect_uri":"http://192.168.1.30:91/callback","code_verifier":"c3IF77ZOp7zn4BNlcyzLFY5TwDcHzKbDIEwson6eoCTMMY9uUw3HJ6O5IsGbQMjAMcTvN9iBDkQrz5JAk46VxHgxg3g0wERc70srLvn5NMtiysgqEOoee5PeOkL9gUxu","code":"def50200f69bf43a01b5dad250c34c1b0f1f1a41507bac7e1f34a76d9d407e342f855f4b0a5eb5f88d9231850dfc373cf10f52d73e50f277b43621f93eabdfe3e2c27155cc73214d1ea6fd916abc2c99689c7578e9fe50c5e27478b81bfaa9e5a10418c952c3c788a8f1dad13fb9d5f12e6d6f31b1fede3cb3efc01847abc809f2023be1ec376726df779fcc466ae99cf4b27d3e43ed368d6216ec3862c1d4d74632f25efffcb102bd41757dde75e6312e72ffc7678d027493924654fecd6a3760b48c307f814af1d34fde90c8ab23a933ec44cddd25d8045aab7f5bd33aafa5b0285ce448edaae25d95ea0be433c5b908f4d72356eb9f40dead08a5627c1fb4a82a40e768dbadd8f78b04ba8001f098bac0eec43cb9ff646a116cb5555b33a9dce75f374cedbaef1d647398d6a0ef491854fb29eb0d00f303dca53df607a7e1b87c52e1b1e7effd19bb35c1a9bbc67bbabea838c4fabcf95828f0db7c7d8e289c1f539427649ddae4cd86667a155459aaf483aa1f9ee48a4f6b93ffff613383049c448d6e7b8e86a457"}} 
[2023-08-18 16:22:49] local.DEBUG: Will contact "http://192.168.1.30:90/oauth/token" for a token.  
127.0.0.1 - - [18/Aug/2023:16:23:14 -0600] "GET /health HTTP/1.1" 200 1105 "-" "Firefly III Health Checker/2.0"
[2023-08-18 16:23:20] local.ERROR: cURL error 28: Connection timed out after 31410 milliseconds (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://192.168.1.30:90/oauth/token {"exception":"[object] (GuzzleHttp\\Exception\\ConnectException(code: 0): cURL error 28: Connection timed out after 31410 milliseconds (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://192.168.1.30:90/oauth/token at /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:210)
[stacktrace]
#0 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(158): GuzzleHttp\\Handler\\CurlFactory::createRejection()
#1 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(110): GuzzleHttp\\Handler\\CurlFactory::finishError()
#2 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlHandler.php(47): GuzzleHttp\\Handler\\CurlFactory::finish()
#3 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(28): GuzzleHttp\\Handler\\CurlHandler->__invoke()
#4 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(48): GuzzleHttp\\Handler\\Proxy::GuzzleHttp\\Handler\\{closure}()
#5 /var/www/html/vendor/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php(64): GuzzleHttp\\Handler\\Proxy::GuzzleHttp\\Handler\\{closure}()
#6 /var/www/html/vendor/guzzlehttp/guzzle/src/Middleware.php(31): GuzzleHttp\\PrepareBodyMiddleware->__invoke()
#7 /var/www/html/vendor/guzzlehttp/guzzle/src/RedirectMiddleware.php(71): GuzzleHttp\\Middleware::GuzzleHttp\\{closure}()
#8 /var/www/html/vendor/guzzlehttp/guzzle/src/Middleware.php(66): GuzzleHttp\\RedirectMiddleware->__invoke()
#9 /var/www/html/vendor/guzzlehttp/guzzle/src/HandlerStack.php(75): GuzzleHttp\\Middleware::GuzzleHttp\\{closure}()
#10 /var/www/html/vendor/guzzlehttp/guzzle/src/Client.php(333): GuzzleHttp\\HandlerStack->__invoke()
#11 /var/www/html/vendor/guzzlehttp/guzzle/src/Client.php(169): GuzzleHttp\\Client->transfer()
#12 /var/www/html/vendor/guzzlehttp/guzzle/src/Client.php(189): GuzzleHttp\\Client->requestAsync()
#13 /var/www/html/vendor/guzzlehttp/guzzle/src/ClientTrait.php(95): GuzzleHttp\\Client->request()
#14 /var/www/html/app/Http/Controllers/TokenController.php(96): GuzzleHttp\\Client->post()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): App\\Http\\Controllers\\TokenController->callback()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(43): Illuminate\\Routing\\Controller->callAction()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(259): Illuminate\\Routing\\ControllerDispatcher->dispatch()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(205): Illuminate\\Routing\\Route->runController()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(799): Illuminate\\Routing\\Route->run()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Session\\Middleware\\StartSession->handle()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(798): Illuminate\\Pipeline\\Pipeline->then()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(777): Illuminate\\Routing\\Router->runRouteWithinStack()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(741): Illuminate\\Routing\\Router->runRoute()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(730): Illuminate\\Routing\\Router->dispatchToRoute()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(200): Illuminate\\Routing\\Router->dispatch()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#45 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#46 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#47 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#48 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#49 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#50 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#51 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#52 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Http\\Middleware\\HandleCors->handle()
#53 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#54 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Http\\Middleware\\TrustProxies->handle()
#55 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#56 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(175): Illuminate\\Pipeline\\Pipeline->then()
#57 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(144): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#58 /var/www/html/public/index.php(76): Illuminate\\Foundation\\Http\\Kernel->handle()
#59 {main}
"} 
192.168.1.238 - - [18/Aug/2023:16:22:49 -0600] "GET /callback?code=def50200f69bf43a01b5dad250c34c1b0f1f1a41507bac7e1f34a76d9d407e342f855f4b0a5eb5f88d9231850dfc373cf10f52d73e50f277b43621f93eabdfe3e2c27155cc73214d1ea6fd916abc2c99689c7578e9fe50c5e27478b81bfaa9e5a10418c952c3c788a8f1dad13fb9d5f12e6d6f31b1fede3cb3efc01847abc809f2023be1ec376726df779fcc466ae99cf4b27d3e43ed368d6216ec3862c1d4d74632f25efffcb102bd41757dde75e6312e72ffc7678d027493924654fecd6a3760b48c307f814af1d34fde90c8ab23a933ec44cddd25d8045aab7f5bd33aafa5b0285ce448edaae25d95ea0be433c5b908f4d72356eb9f40dead08a5627c1fb4a82a40e768dbadd8f78b04ba8001f098bac0eec43cb9ff646a116cb5555b33a9dce75f374cedbaef1d647398d6a0ef491854fb29eb0d00f303dca53df607a7e1b87c52e1b1e7effd19bb35c1a9bbc67bbabea838c4fabcf95828f0db7c7d8e289c1f539427649ddae4cd86667a155459aaf483aa1f9ee48a4f6b93ffff613383049c448d6e7b8e86a457&state=FrAHwauL24rOq2QHpToWt5HX6puqNNKrCIYRbR7d HTTP/1.1" 500 7730 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0"

**Comment by ghost** [comment, marked as answer]
Ahh, I figured it out. I hadn't allowed 90/tcp and 91/tcp through the server firewall. As soon as I did that and reloaded the firewall, I got to the home screen of the importer.

Apologies, I read all of the guidelines before submitting this but it was an issue with me after all.

**Reply by JC5** [reply]
Glad to hear you found it!
